app/src/main/kotlin/com/wire/android/mapper/RegularMessageContentMapper.kt

```diff
@@ -158,8 +158,12 @@ class RegularMessageMapper @Inject constructor(
     fun toUIMessageContent(assetMessageContentMetadata: AssetMessageContentMetadata, message: Message, sender: User?): UIMessageContent =
         with(assetMessageContentMetadata.assetMessageContent) {
             when {
-                assetMessageContentMetadata.isDisplayableImage() && !assetMessageContentMetadata.assetMessageContent.hasValidRemoteData() ->
-                    UIMessageContent.PreviewAssetMessage
+                assetMessageContentMetadata.isDisplayableImage()
+                        // assets uploaded by other clients have upload status NOT_UPLOADED
+                        && assetMessageContentMetadata.assetMessageContent.uploadStatus == Message.UploadStatus.NOT_UPLOADED
+                        // sometimes we can receive two asset messages, we want to show the image only after we get all required data
+                        && !assetMessageContentMetadata.assetMessageContent.hasValidRemoteData() ->
```

Comment:
perhaps it would be better for have this inside assetMessageContentMetadata ? I feel like it belongs more to that class instead of this logic hanging loosely inside the mapper

Reply:
Sure, you're right, I moved it there